# Online Retails Dataset

![CHD](retail.jpg)

This dataset contains all purchases made for an `online retail company based in the UK` during an eight month period.

The size of this dataset is about `(541909, 8)` rows and columns (invoice number, Stock code, Description, Quantity, Invoice date, Unit price, Customer ID and Country)

## Apriori Algorithm is used

Apriori algorithm was the first algorithm that was proposed for frequent itemset mining. It was later improved by R Agarwal and R Srikant and came to be known as Apriori. This algorithm uses two steps “join” and “prune” to reduce the search space. It is an iterative approach to discover the most frequent itemsets.

In [1]:
#pip install mlxtend

### Association Rules

In [2]:
## Here I have imported essential libraries
import numpy as np
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [3]:
## import the dataset into Pandas Dataframe. The head() function is used to get the first n rows.
df = pd.read_csv("Online Retail.csv")
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

       InvoiceDate  UnitPrice  CustomerID         Country  
0  01/12/2010 8:26       2.55     17850.0  United Kingdom  
1  01/12/2010 8:26       3.39     17850.0  United Kingdom  
2  01/12/2010 8:26       2.75     17850.0  United Kingdom  
3  01/12/2010 8:26       3.39     17850.0  United Kingdom  
4  01/12/2010 8:26       3.39     17850.0  United Kingdom

In [4]:
df.shape

(541909, 8)

### Data Preparation

In [5]:
#Data Cleaning
df['Description'] = df['Description'].str.strip() #removes spaces from beginning and end of the description column
df.dropna(axis=0, subset=['InvoiceNo'], inplace=True) #removes duplicate invoice
df['InvoiceNo'] = df['InvoiceNo'].astype('str') #converting invoice number to be type string, needed as the algorithm understand the invoice number as a string data.
df = df[~df['InvoiceNo'].str.contains('C')] #remove the credit transactions,a credit type of invoice is removed with alphabet 'c'
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

       InvoiceDate  UnitPrice  CustomerID         Country  
0  01/12/2010 8:26       2.55     17850.0  United Kingdom  
1  01/12/2010 8:26       3.39     17850.0  United Kingdom  
2  01/12/2010 8:26       2.75     17850.0  United Kingdom  
3  01/12/2010 8:26       3.39     17850.0  United Kingdom  
4  01/12/2010 8:26       3.39     17850.0  United Kingdom

In [6]:
df['Country'].value_counts() ## the value count or record count for different countries

United Kingdom          487622
Germany                   9042
France                    8408
EIRE                      7894
Spain                     2485
Netherlands               2363
Belgium                   2031
Switzerland               1967
Portugal                  1501
Australia                 1185
Norway                    1072
Italy                      758
Channel Islands            748
Finland                    685
Cyprus                     614
Sweden                     451
Unspecified                446
Austria                    398
Denmark                    380
Poland                     330
Japan                      321
Israel                     295
Hong Kong                  284
Singapore                  222
Iceland                    182
USA                        179
Canada                     151
Greece                     145
Malta                      112
United Arab Emirates        68
European Community          60
RSA                         58
Lebanon 

In [7]:
df.shape

(532621, 8)

I am filtering the data on one particular country Separating transactions for Germany

I have  taken Invoice number, description and sum of the quantity.

The entire code will give me a basket of transaction.

In [8]:
mybasket = (df[df['Country'] =="Germany"]
          .groupby(['InvoiceNo', 'Description'])['Quantity'] 
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))


### Viewing the transaction basket
In my basket of Germany there are number of products shown as columns 

The value given under a product column is 0 example 10 colour spaceboy pen under invoice number 536527 is 0.0 that means that product was not present in that invoice.

If the number under a product column is 123... that means that product was the part of that particular invoice that many times.

In [9]:
mybasket.head()

Description  10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
InvoiceNo                                                         
536527                          0.0                         0.0   
536840                          0.0                         0.0   
536861                          0.0                         0.0   
536967                          0.0                         0.0   
536983                          0.0                         0.0   

Description  12 IVORY ROSE PEG PLACE SETTINGS  \
InvoiceNo                                       
536527                                    0.0   
536840                                    0.0   
536861                                    0.0   
536967                                    0.0   
536983                                    0.0   

Description  12 MESSAGE CARDS WITH ENVELOPES  12 PENCIL SMALL TUBE WOODLAND  \
InvoiceNo                                                                     
536527                                   0.0                            0.0   
536840                                   0.0                            0.0   
536861                                   0.0                            0.0   
536967                                   0.0                            0.0   
536983                                   0.0                            0.0   

Description  12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS SMALL TUBE SKULL  \
InvoiceNo                                                                       
536527                                       0.0                          0.0   
536840                                       0.0                          0.0   
536861                                       0.0                          0.0   
536967                                       0.0                          0.0   
536983                                       0.0                          0.0   

Description  12 PENCILS TALL TUBE POSY  12 PENCILS TALL TUBE RED RETROSPOT  \
InvoiceNo                                                                    
536527                             0.0                                 0.0   
536840                             0.0                                 0.0   
536861                             0.0                                 0.0   
536967                             0.0                                 0.0   
536983                             0.0                                 0.0   

Description  12 PENCILS TALL TUBE SKULLS  ...  YULETIDE IMAGES GIFT WRAP SET  \
InvoiceNo                                 ...                                  
536527                               0.0  ...                            0.0   
536840                               0.0  ...                            0.0   
536861                               0.0  ...                            0.0   
536967                               0.0  ...                            0.0   
536983                               0.0  ...                            0.0   

Description  ZINC  HEART T-LIGHT HOLDER  ZINC  STAR T-LIGHT HOLDER  \
InvoiceNo                                                            
536527                              0.0                        0.0   
536840                              0.0                        0.0   
536861                              0.0                        0.0   
536967                              0.0                        0.0   
536983                              0.0                        0.0   

Description  ZINC BOX SIGN HOME  ZINC FOLKART SLEIGH BELLS  \
InvoiceNo                                                    
536527                      0.0                        0.0   
536840                      0.0                        0.0   
536861                      0.0                        0.0   
536967                      0.0                        0.0   
536983                      0.0                        0.0   

Description  ZINC HEART LATTIC

Converting all positive vaues (all other numbers like 1,2,3 etc) to 1 and others to 0.That is what my algorithm expects as input.

This is what my algorithm of association analysis expects its input data to be in the form of 0 and 1.

In [10]:
def my_encode_units(x):
    if x <= 0:              # all the numbers less than 0 to 0
        return 0 
    if x >= 1:              # all the numbers greater than 1 to 1
        return 1

my_basket_sets = mybasket.applymap(my_encode_units)
my_basket_sets.drop('POSTAGE', inplace=True, axis=1) #Removed "postage" as an item , it is not actually required.

In [11]:
my_basket_sets

Description  10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
InvoiceNo                                                         
536527                            0                           0   
536840                            0                           0   
536861                            0                           0   
536967                            0                           0   
536983                            0                           0   
...                             ...                         ...   
581266                            0                           0   
581494                            0                           0   
581570                            0                           0   
581574                            0                           0   
581578                            0                           0   

Description  12 IVORY ROSE PEG PLACE SETTINGS  \
InvoiceNo                                       
536527                                      0   
536840                                      0   
536861                                      0   
536967                                      0   
536983                                      0   
...                                       ...   
581266                                      0   
581494                                      0   
581570                                      0   
581574                                      0   
581578                                      0   

Description  12 MESSAGE CARDS WITH ENVELOPES  12 PENCIL SMALL TUBE WOODLAND  \
InvoiceNo                                                                     
536527                                     0                              0   
536840                                     0                              0   
536861                                     0                              0   
536967                                     0                              0   
536983                                     0                              0   
...                                      ...                            ...   
581266                                     0                              0   
581494                                     0                              0   
581570                                     0                              0   
581574                                     0                              0   
581578                                     0                              0   

Description  12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS SMALL TUBE SKULL  \
InvoiceNo                                                                       
536527                                         0                            0   
536840                                         0                            0   
536861                                         0                            0   
536967                                         0                            0   
536983                                         0                            0   
...                                          ...                          ...   
581266                                         0                            0   
581494                                         0                            0   
581570                                         0                            0   
581574                                         0                            0   
581578                                         0                            0   

Description  12 PENCILS TALL TUBE POSY  12 PENCILS TALL TUBE RED RETROSPOT  \
InvoiceNo                                                                    
536527                               0                                   0   
536840                               0                                   0   
536861                               0                                   0   
536967            

### Training Model

Learning purpose: If the number of transactions is 5, your minimum support count would be 5*60/100 = 3.



## Apriori Algorithm 

Generating frequent itemsets.I am running Apriori on my_basket_sets which I have imported from mlxtend.frequent_patterns(in the first cell)

In [12]:
my_frequent_itemsets = apriori(my_basket_sets, min_support=0.07, use_colnames=True)

I'm taking my_frequent_itemset as input and out of support lift and confidence I have taken the metric as lift

In [13]:
#generating rules
my_rules = association_rules(my_frequent_itemsets, metric="lift", min_threshold=1) 

In [14]:
#viewing top 100 rules
my_rules.head(100)

antecedents                           consequents  \
0  (ROUND SNACK BOXES SET OF4 WOODLAND)    (PLASTERS IN TIN WOODLAND ANIMALS)   
1    (PLASTERS IN TIN WOODLAND ANIMALS)  (ROUND SNACK BOXES SET OF4 WOODLAND)   
2   (ROUND SNACK BOXES SET OF 4 FRUITS)  (ROUND SNACK BOXES SET OF4 WOODLAND)   
3  (ROUND SNACK BOXES SET OF4 WOODLAND)   (ROUND SNACK BOXES SET OF 4 FRUITS)   
4  (ROUND SNACK BOXES SET OF4 WOODLAND)                  (SPACEBOY LUNCH BOX)   
5                  (SPACEBOY LUNCH BOX)  (ROUND SNACK BOXES SET OF4 WOODLAND)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.245077            0.137856  0.074398    0.303571  2.202098   
1            0.137856            0.245077  0.074398    0.539683  2.202098   
2            0.157549            0.245077  0.131291    0.833333  3.400298   
3            0.245077            0.157549  0.131291    0.535714  3.400298   
4            0.245077            0.102845  0.070022    0.285714  2.778116   
5            0.102845            0.245077  0.070022    0.680851  2.778116   

   leverage  conviction  
0  0.040613    1.237951  
1  0.040613    1.640006  
2  0.092679    4.529540  
3  0.092679    1.814509  
4  0.044817    1.256018  
5  0.044817    2.365427

#### Product A 'Round Snack Boxes set  of 4 Woodland' --> Product B 'Plasters in tin Woodland Animals' 
#### Support = 0.074398
#### Confidence = 0.303571
#### Lift = 2.202098

### Making Recommendations 
#### how do we use the above rules to make recomendations?

In [15]:
## Consider the rule table above RULE NUMBER #5 people brought item B 112 times 
my_basket_sets['ROUND SNACK BOXES SET OF4 WOODLAND'].sum()

112

In [16]:
## Consider the rule table above RULE NUMBER #5 people brought item A 47 times 
my_basket_sets['SPACEBOY LUNCH BOX'].sum()

47

#### That means out people who are buying the item 'ROUND SNACK BOXES SET OF4 WOODLAND' 112 times out of which only 47 times the item SPACEBOY LUNCH BOX is being brought.
#### Therefore the rule says that there is a good support, confidence and lift for these combinations so we can recommend SPACEBOY LUNCH BOX to someone who is buying ROUND SNACK BOXES SET OF4 WOODLAND.

In [17]:
#Filtering rules based on condition
my_rules[ (my_rules['lift'] >= 3) &
       (my_rules['confidence'] >= 0.3) ]

antecedents                           consequents  \
2   (ROUND SNACK BOXES SET OF 4 FRUITS)  (ROUND SNACK BOXES SET OF4 WOODLAND)   
3  (ROUND SNACK BOXES SET OF4 WOODLAND)   (ROUND SNACK BOXES SET OF 4 FRUITS)   

   antecedent support  consequent support   support  confidence      lift  \
2            0.157549            0.245077  0.131291    0.833333  3.400298   
3            0.245077            0.157549  0.131291    0.535714  3.400298   

   leverage  conviction  
2  0.092679    4.529540  
3  0.092679    1.814509